barretenberg/cpp/src/barretenberg/vm2/tracegen/execution_trace.cpp

```diff
@@ -67,21 +69,135 @@ constexpr std::array<Column, NUM_REGISTERS> REGISTER_MEM_OP_COLUMNS = {
     C::execution_mem_op_4_, C::execution_mem_op_5_, C::execution_mem_op_6_,
 };
 
+/**
+ * @brief Helper struct to track info after "discard" preprocessing.
+ */
+struct ErroringContexts {
+    bool app_logic_error = false;
+    bool teardown_error = false;
+    uint32_t app_logic_exit_context_id = 0;
+    uint32_t teardown_exit_context_id = 0;
+    std::unordered_map<uint32_t, bool> does_context_error;
+};
+
+/**
+ * @brief Preprocess execution events to determine which contexts will error.
+ *
+ * @details This is used during trace-generation to populate the `discard` and `dying_context_id` columns
+ * which must be set throughout a context that will EVENTUALLY error. So we need to do a
+ * preprocessing pass so that we can set these columns properly during trace-generation for rows
+ * in a dying context before the the actual error event is reached.
+ *
+ * @param ex_events The execution events.
+ * @return The erroring contexts.
+ */
+ErroringContexts preprocess_for_discard(
+    const simulation::EventEmitterInterface<simulation::ExecutionEvent>::Container& ex_events)
+{
+    ErroringContexts dying_info;
+
+    // Preprocessing pass 1: find the events that exit the app logic and teardown phases
+    for (const auto& ex_event : ex_events) {
+        bool is_exit = ex_event.wire_instruction.get_exec_opcode() == ExecutionOpCode::RETURN ||
```

Comment:
`get_exec_opcode()` will crash if instruction fetching failed, you need an extra guard here. You can know if that happened using `ExecutionError error`. Also each call accesses a map so if you can cache the result, better :)